src/layout/node.ts

```diff
@@ -17,29 +17,44 @@ export default function(node: Node, source: Source): void {
     }
 
     let call = dom.has(node) ? "update" : "add";
+    let parent = node.parentElement ? node.parentElement : null;
     switch (node.nodeType) {
         case Node.DOCUMENT_TYPE_NODE:
             let doctype = node as DocumentType;
             let docAttributes = { name: doctype.name, publicId: doctype.publicId, systemId: doctype.systemId };
             let docData = { tag: "*D", attributes: docAttributes };
-            dom[call](node, docData, source);
+            dom[call](node, parent, docData, source);
+            break;
+        case Node.DOCUMENT_FRAGMENT_NODE:
+            let shadowRoot = (node as ShadowRoot);
+            if (shadowRoot.host) {
+                // See: https://wicg.github.io/construct-stylesheets/
+                // For more details on adoptedStyleSheets for shadow dom
+                let style = "";
+                let adoptedStyleSheets: CSSStyleSheet[] = "adoptedStyleSheets" in shadowRoot ? shadowRoot["adoptedStyleSheets"] : [];
+                for (let styleSheet of adoptedStyleSheets) { style += getCssRules(styleSheet); }
+                let fragementData = { tag: "*S", attributes: { style } };
```

Comment:
>tag [](start = 38, length = 3)

Are these tags sent over the wire or only used internally?

Reply:
Sent over the wire